### ISSUE TITLE
Add ability for transprots to pick up their cargo

### ISSUE THREAD

**Issue by dragunoff** [OpenRA/OpenRA#14108]
Right now cargo/passengers can be ordered to go into transports but transports can not be ordered to pick up their cargo/passengers. Such feature exists in other RTS (Starcraft for example) and can be very useful. 

Imagine how much easier it will be to pick up stranded engineers (or Tanya). Now you have to first move the transport to where the passenger is and then order the passenger in (if you don't forget that you sent that transport in the first place). It can be extra useful if combined with queued orders - drive to the passenger, pick up, drive off to somewhere else with the passenger on board.

**Comment by KOYK** [comment]
Warcraft 3 also.

**Comment by dragunoff** [comment]
This might be difficult to pull off for naval transports because the cargo but could be in a position that makes it impossible to pick up.

**Comment by dragunoff** [comment]
I have coded a proof of concept:

https://user-images.githubusercontent.com/1355810/103307063-d436da80-4a17-11eb-8266-b1af72988a2c.mp4

It raises some UI questions:
- Should this be the default action when targeting passengers? Probably not as it's common to give move orders over friendly units. 
- Should this be attached to a modifier key then? Attaching to <kbd>Ctrl</kbd> means that transports would not be able to force fire friendly passengers. Attaching to <kbd>Alt</kbd> seems better though we then lose force moving over friendly units (which may be a valid scenario).
- Should this receive it's own hotkey? Would that hotkey be attached to a button? Where would that button live?  :thinking: 

In all cases we bump into some of the limitations of the CNC UI. This is one of the features that calls for something like unit abilities bar. A unit abilities bar can also solve (or at least ease) many of the UI issues with Chronotanks, minelayers and aircraft :thinking: This has popped up in several other discussions and I should probably file a separate issue to describe the idea.

**Comment by pchote** [comment]
I'm strongly against having a unit-abilities UI. This takes the game too far from the "Classic C&C" feel, and isn't needed to make a good/popular game - just a different game. Consider that the most popular C&C games remain RA2, RA1, and the remastered editions, none of which had this.

**Comment by dragunoff** [comment]
I agree that an abilities bar isn't needed to make the game good (the games are quite good already). And I agree that this would be a departure from that "classic C&C feel". Even though OpenRA's mods are already quite different games to their original counterparts :man_shrugging: I'm not pushing for it but I think it's an idea worth exploring. This case is also a reminder of just how hard it is to get new features accepted upstream because of the fine line that the default mods need to walk. Perhaps we need a major non-cnc mod (downstream) where such ideas can be developed :thinking:

**Comment by deleted-user-1** [comment]
> Perhaps we need a major non-cnc mod (downstream) where such ideas can be developed

Imagine a non cnc upstream mod (literally, all you need is a universe, lore, storyline, concept art, artists and core mechanics etc.). OpenRA's "backport cnc3"-dogma applies only to cnc content.

Also related to #18486 - IMO it's an antifeature and the existing mechanics are sufficient (but could be improved on the activity level as described there).

**Comment by PunkPun** [comment]
> * Should this be the default action when targeting passengers? Probably not as it's common to give move orders over friendly units.

It also dends on the control cheme. With left mouse orders you can't give move orders over friedly units, if you try you will just select those units instead. You need to use force move

**Comment by KOYK** [comment]
> I agree that an abilities bar isn't needed to make the game good (the games are quite good already). And I agree that this would be a departure from that "classic C&C feel". Even though OpenRA's mods are already quite different games to their original counterparts  I'm not pushing for it but I think it's an idea worth exploring. This case is also a reminder of just how hard it is to get new features accepted upstream because of the fine line that the default mods need to walk. Perhaps we need a major non-cnc mod (downstream) where such ideas can be developed

Can we have this for none official mods? it will be like a gold mine to us.

to add to that, for a long time I was trying to figure a way on how to mod the command bar to do it, but I dont thing its possible, and even if it is well its impossible to pull it off, so ive ended up using a hackish way to do it, for buildings, but for units aint happening. having ability buttons like CNC3 would solve so many issues in the modding community.

**Comment by dragunoff** [comment]
I tinkered a little bit more with this feature and figured that doing it properly requires quite a bit of changes to the way transports and passengers interact. So before I go any further I'd like to know whether this fits with the classic CnC gameplay :thinking: To me it's a straight-up improvement that doesn't depart from CnC but I'd like to hear what others think.

**Comment by pchote** [comment]
What does "doing it properly" and "quite a bit of changes" mean here? I'm hesitant that this may end up creating new / more problems than it solves (e.g. breaking expected cursor behaviours), but don't have an objection if it works well without requiring any fundamental UI changes.

**Comment by dragunoff** [comment]
@pchote I meant it requires changes to the activities because right now everything is happening in `RideTransport` from the passenger's perspective. I think it has to be shifted to the transport in a new `PickupPassenger` activity, kind of like how `PickupUnit` works for carryalls. The UI would remain the same - the only addition would be a new unit target order generator for transports.

**Comment by pchote** [comment]
What happens if you have a naval transport selected and click on a unit that is several cells away from the beach?

**Comment by dragunoff** [comment]
> What happens if you have a naval transport selected and click on a unit that is several cells away from the beach?

Nothing, both units just stand still. That has to be worked out of course. Perhaps the transport should move to the shore and then the passenger should move toward it. Or this feature may even be disabled for naval transports.

**Comment by deleted-user-1** [comment]
> Imagine how much easier it will be to pick up stranded engineers (or Tanya). Now you have to first move the transport to where the passenger is and then order the passenger in (if you don't forget that you sent that transport in the first place). It can be extra useful if combined with queued orders - drive to the passenger, pick up, drive off to somewhere else with the passenger on board.

I thought about this again and would like to explain again why I'm niot a fan of the idea for default OpenRA mods:
- These are really some of the last bits where a player with good multitasking can shine because it's a bit harder to use transports of any kind. The current system awards using control groups and requires you to split your attention. Something I consider good for rts.
- There are annoying parts that I already described could be fixed and make loading faster when selecting many units.
- For default mods, there are no cursors available (unlike in TS for the carryall).
- Might cause the exceptions you already noted (may even be disabled for naval transports).

So +1 for the engine support, not really convinced it's good for default mods.

**Comment by Smittytron** [comment]
Related #18486
